CsharpExtras/Extensions/ArrayExtension.cs

```diff
@@ -228,5 +229,100 @@ public static void AssertIndexIsInBounds<TVal>(this TVal[] array, int index)
                     "Should be in the range [{1}, {2}]", 0, length));
             }
         }
+
+        /// <summary>
+        /// Converts this array to a reverse list-based dictionary
+        /// </summary>
+        public static IListValuedDictionary<TVal, int> ConvertToReverseListValuedDictionary<TVal>(this TVal[] array, int indexOffset)

```

Comment:
I think it makes sense to add overrides for these functions that default the offset to 0. I think it's great that you support it, but I don't see many cases where they would be used (aside from one-based arrays, but they probably should implement a version of this that automatically sets the offset to 1 to avoid confusion).